docs/migrations-guidelines.md

```diff
@@ -0,0 +1,85 @@
+# Migration and Testing Guidelines
+
+## Overview
+
+This document outlines best practices and guidelines for writing migration scripts and their corresponding test cases.
+
+The focus is on ensuring data integrity, handling errors gracefully, and maintaining consistency across mobile app versions.
+
+## Migration Guidelines
+
+1. **Pre-Validation Checks**:
+
+- Ensure the state to be migrated is valid before running the migration logic,
+- Verify the version and integrity of the state with utility functions like `ensureValidState`.
+
+2. **Error Handling**:
+
+The following prevents manipulation of potentially corrupted data, enforcing corrective action before proceeding.
+
+- Log erros with `captureException` from `@sentry/react-native`, which is crucial for diagnosing issues post-migration,
+- Ensure that error messages are descriptive: include the migration number and a clear description of the issue,
+- If an exception is detected, indicating potential data corruption, halt the migration process and return the intial state,
+
+3. **State Integrity Checks**:
+
+- Checks the state's structure and types carefully,
+- Validate the state and its nested properties using functions like `isObject` and `hasProperty`,
+- Prevent data corruption by halting the migration on any inconsistencies and logging errors.
+
+5. **Avoid Type Casting**:
+
+- Do not use type casting (for example, `as` keyword in TypeScript) to manipulate the state or its properties,
+- Type casting can mask underlying data structure issues and make the code less resilient in case of change in dependencies or data structures,
+- Instead, use type guards and runtime checks to ensure data integrity and compatibility.
+
+6. **Return State**:
+
+- Always return the state at the end of the migration function, whether it was modified or not,
+- Returning the state ensures that the migration process completes and the state is passed to the next migrations.
+
+## Testing Guidelines
+
+1. **Mocking**:
+
+- Mock dependencies with `jest.mock` (for example `@sentry/react-native`),
+- Mocking isolates the migration under test and prevents side effects.
+
+2. **Initial State Setup**:
+
+- Create an initial state that reflects possible real-world scenarios, including edge cases,
+- if needed, create multiple initial states and use them each in a test for this specific case,
+- Use utilities like `merge` from `lodash` to combine base states with specific test case modifications.
+
+4. **Invalid State Scenarios**:
```

Comment:
I realized that we don't currently handle invalid states gracefully. Let's expand on this and mention to backfill data that might be missing or corrupted. For example, we shouldn't just return original state since it will most likely lead to a crash at some point in the app.

Reply:
Hmm but with that we might be not able to fix the original issue, and to do that I'm imagining that we need to return the corrupted data to some default state, that might delete data for the fix we need

Food for thoughts

Reply:
It won't be perfect right now since we don't have a standardized reconciliation mechanism but to your point, we should be able to fallback to a default state or at least derive it from the original state. Worst case scenario would be to return original state but as we've seen in the past, that almost always leads to unexpected conditions and/or crash.

Reply:
I'm not sure if this would work with any controllers data! Will reach out for we to schedule 15 minutes to work on an example

Reply:
Falling back to default state sounds worse than letting the app crash, because we could erase user data that we can't get back. That's why we started this pattern in the extension migrations

Reply:
We could consider using a default state fallback and stashing the invalid state somewhere for recovery later. Some risk of bloating state size there, but as long as our default state is small and we don't keep backups too long, that should be OK.

Reply:
Makes sense, we'll have to think about this some more before we provide a good solution for gracefully handling. We can leave this out for now and continue brainstorming. @tommasini